src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -883,138 +883,110 @@ ucs_status_t uct_dc_mlx5_iface_tag_recv_cancel(uct_iface_h tl_iface,
     uct_dc_fc_sender_data_t sender;
-    uct_dc_fc_request_t *dc_req;
-    struct mlx5_wqe_av mlx5_av;
     uct_ib_mlx5_base_av_t av;
-    ucs_status_t status;
-    uintptr_t sender_ep;
+    struct mlx5_wqe_av mlx5_av;
     struct ibv_ah *ah;
-    ucs_time_t now;
-    khiter_t it;
-    int ret;
+    ucs_status_t status;
 
     UCT_DC_MLX5_TXQP_DECL(txqp, txwq);
 
     ucs_assert((sizeof(uint8_t) + sizeof(sender_ep)) <=
-                UCT_IB_MLX5_AV_FULL_SIZE);
+                       UCT_IB_MLX5_AV_FULL_SIZE);
 
-    UCT_DC_MLX5_CHECK_DCI_RES(iface, dc_ep);
-    UCT_DC_MLX5_IFACE_TXQP_GET(iface, dc_ep, txqp, txwq);
+    /* TODO: look at common code with uct_ud_mlx5_iface_get_av */
+    if (fc_req->sender.payload.is_global) {
+        uct_ib_iface_fill_ah_attr_from_gid_lid(
+                ib_iface, fc_req->lid,
+                ucs_unaligned_ptr(&fc_req->sender.payload.gid),
+                iface->super.super.super.gid_info.gid_index, 0, &ah_attr);
 
-    dc_req = ucs_derived_of(req, uct_dc_fc_request_t);
+        status = uct_ib_iface_create_ah(ib_iface, &ah_attr, &ah);
+        if (status != UCS_OK) {
+            return status;
+        }
 
-    if (op == UCT_RC_EP_FC_PURE_GRANT) {
-        ucs_assert(req != NULL);
+        uct_ib_mlx5_get_av(ah, &mlx5_av);
+    }
 
-        sender_ep = (uintptr_t)dc_req->sender.ep;
+    /* Note: AV initialization is copied from exp verbs */
+    av.stat_rate_sl = ib_iface->config.sl; /* (attr->static_rate << 4) |
+                                              attr->sl */
+    av.fl_mlid      = ib_iface->path_bits[0] & 0x7f;
 
-        /* TODO: look at common code with uct_ud_mlx5_iface_get_av */
-        if (dc_req->sender.payload.is_global) {
-            uct_ib_iface_fill_ah_attr_from_gid_lid(ib_iface, dc_req->lid,
-                                                   ucs_unaligned_ptr(&dc_req->sender.payload.gid),
-                                                   iface->super.super.super.gid_info.gid_index,
-                                                   0, &ah_attr);
+    /* lid in fc_req is in BE already  */
+    if (uct_ib_iface_is_roce(ib_iface)) {
+        av.rlid = htons(UCT_IB_ROCE_UDP_SRC_PORT_BASE);
+    } else {
+        av.rlid = fc_req->lid | htons(ib_iface->path_bits[0]);
+    }
 
-            status = uct_ib_iface_create_ah(ib_iface, &ah_attr, &ah);
-            if (status != UCS_OK) {
-                goto out_dci_put;
-            }
+    av.dqp_dct = htonl(fc_req->dct_num);
 
-            uct_ib_mlx5_get_av(ah, &mlx5_av);
-        }
+    if (!iface->ud_common.config.compact_av || ah_attr.is_global) {
+        av.dqp_dct |= UCT_IB_MLX5_EXTENDED_UD_AV;
+    }
 
-        /* Note av initialization is copied from exp verbs */
-        av.stat_rate_sl = ib_iface->config.sl; /* (attr->static_rate << 4) | attr->sl */
-        av.fl_mlid      = ib_iface->path_bits[0] & 0x7f;
+    UCT_DC_MLX5_CHECK_DCI_RES(iface, ep);
+    UCT_DC_MLX5_IFACE_TXQP_GET(iface, ep, txqp, txwq);
```

Comment:
Maybe better take dci at the beginning of func and release it later if some error occured.
The case when there is no dci may happen quite often (and repeatedly), so with current approach we will always do extra  work (create AH, etc) for no reason

Reply:
make sense, done